main.py

```diff
@@ -132,60 +134,66 @@ def process_players(players: List, limit: int) -> Tuple[List, List]:
 
 
 def get_clean_data(
-    data: Iterator, required_field_and_type: Set[Tuple[str, str]]
-) -> Iterator[Dict]:
+    data: Iterable, required_field_and_type: Set[Tuple[str, type]]
+) -> Iterator[datastore.Entity]:
     for item in data:
         if all(
-            field in item and isinstance(item[field], type)
-            for field, type in required_field_and_type
+            field_name in item and isinstance(item[field_name], type_name)
+            for field_name, type_name in required_field_and_type
         ):
             yield item
 
 
-def get_player_list_name_by_type(player_type: str) -> str:
+def get_player_list_name_by_type(player_type: str) -> Optional[str]:
     if player_type == "double":
         return "double_players"
     elif player_type == "single":
         return "single_players"
+    return None
 
 
-def process_add(group_id: str, player_type: str, player_name: str) -> None:
+def process_add(group_id: str, player_type: str, player_name: str) -> Optional[str]:
     with datastore_client.transaction() as transaction:
         group_entity = fetch_group(group_id, transaction=transaction)
         if not group_entity:
-            return
+            return f"Can't find group with ID {group_id}"
         player_list_name = get_player_list_name_by_type(player_type)
-        if player_name.lower() in {x["name"].lower() for x in group_entity[player_list_name]}:
-            print("Player with the same name already exist!")
+        if player_name.lower() in {
+            x["name"].lower() for x in group_entity[player_list_name]
+        }:
+            return "Player with the same name already exist"
         else:
             group_entity[player_list_name].append(
                 {"name": player_name, "signup_time": datetime.utcnow()}
             )
             datastore_client.put(group_entity)
+            return None
 
 
-def process_remove(group_id: str, player_type: str, player_name: str):
+def process_remove(group_id: str, player_type: str, player_name: str) -> Optional[str]:
     with datastore_client.transaction() as transaction:
         group_entity = fetch_group(group_id, transaction=transaction)
         if not group_entity:
-            return
+            return f"Can't find group with ID {group_id}"
```

Comment:
Come to think of it, feels like `try` and `except` would be better in this method...

Reply:
Wrapped all db calls with try/except in the last commit.